prov/gni/src/gnix_msg.c

```diff
@@ -3136,77 +3189,55 @@ ssize_t _gnix_recvv(struct gnix_fid_ep *ep, const struct iovec *iov,
 			goto pdc_exit;
 		}
 
+		for (i = 0; i < count; i++) {
+			req->msg.recv_info[i].recv_addr = (uint64_t) iov[i].iov_base;
+			req->msg.recv_info[i].recv_len = iov[i].iov_len;
+		}
+
 		if (req->msg.send_flags & GNIX_MSG_RENDEZVOUS) {
 			req->work_fn = __gnix_rndzv_iov_req_build;
-			if (!desc) {	/* Register the memory for the user */
-				for (i = 0; i < count; i++) {
-					auto_mr = NULL;
-					ret = gnix_mr_reg(&ep->domain->
-							  domain_fid.fid,
-							  iov[i].iov_base,
-							  iov[i].iov_len,
-							  FI_READ | FI_WRITE, 0,
-							  0, 0, &auto_mr,
-							  NULL);
-
+			if (!(req->vc->modes & GNIX_VC_MODE_XPMEM)) {
+				if (!desc) {
+					ret = __gnix_msg_register_iov(ep->domain,
+								      iov,
+								      count,
+								      req->msg.recv_md);
 					if (ret != FI_SUCCESS) {
 						GNIX_DEBUG(FI_LOG_EP_DATA,
 							   "Failed to "
 							   "auto-register"
 							   " local buffer: %s\n"
 							   , fi_strerror(-ret));
-
-						for (i--; i >= 0; i--) {
-							fi_close(&req->msg.recv_md[i]->mr_fid.fid);
-						}
-
 						goto err;
 					}
+					req->msg.send_flags |= FI_LOCAL_MR;
```

Comment:
That was in the original code.  We're using that to indicate that the buffer has been registered internally and not by the user.